### PR TITLE
Exclude "distinct from" from hard coded references flag

### PR DESCRIPTION
**Pull request by graciegoheen**
This is a:
- [x] bug fix PR with no breaking changes
- [ ] new functionality

## Link to Issue 
Closes https://github.com/dbt-labs/dbt-project-evaluator/issues/338

## Description & motivation
SQL that uses `is distinct from` or `is not distinct from` is being incorrectly flagged as a hard coded reference. 

Example:
```
    select
        old.* exclude (fact_effective_thru, fact_history_is_current) 
        ,current_timestamp() as fact_effective_thru
        ,false as fact_history_is_current
    from
        {{ref(old_state)}} as old
    inner join {{ref(new_state)}} as new
        using(fact_durable_key)
    where
        old.ALL_SOURCE_ATTRIBUTES is distinct from new.ALL_SOURCE_ATTRIBUTES
```
`new.ALL_SOURCE_ATTRIBUTES` is being flagged.

I added a negative lookback for `distinct ` to all regex checks.

## Integration Test Screenshot
Screenshot 2024-01-04 at 3 55 43 PM

## Checklist
- [ ] I have verified that these changes work locally on the following warehouses (Note: it's okay if you do not have access to all warehouses, this helps us understand what has been covered)
    - [ ] BigQuery
    - [ ] Postgres
    - [ ] Redshift
    - [x] Snowflake
    - [ ] Databricks
    - [ ] DuckDB
    - [ ] Trino/Starburst
- [ ] I have updated the README.md (if applicable)
- [ ] I have added tests & descriptions to my models (and macros if applicable)

### CODE REVIEW

**Comment by b-per** [comment]
Yeahh, regexes to review!

File: integration_tests/models/staging/stg_model_5.sql [hunk 1]
```diff
@@ -1 +1,3 @@
-{{ ref('fct_model_9') }}
\ No newline at end of file
+select 
+    * 
+from {{ ref('fct_model_9') }}
```
Comment (dave-connors-3) [inline integration_tests/models/staging/stg_model_5.sql:3]: was this edit needed?

**Comment by dave-connors-3** [comment]
@graciegoheen did you make changes to the seeds at all to catch this new model?

**Comment by graciegoheen** [comment]
@dave-connors-3 No just added this to `int_model_5`:
```
select 1 as id
-- from {{ ref('int_model_4') }}
-- inner join {{ ref('stg_model_1') }}
-- on int_model_4.join_field = stg_model_1.join_field
-- where int_model_4.id is distinct from stg_model_1.id
```

**Comment by dave-connors-3** [comment]
@graciegoheen ah i see i guess we would indeed _not_ expect a new violation there lol

**Review by dave-connors-3 — APPROVED** [review]

File: integration_tests/models/staging/stg_model_5.sql (hunk 1 above)
Comment (graciegoheen) [reply integration_tests/models/staging/stg_model_5.sql:3]: no lol

Comment (dave-connors-3) [reply integration_tests/models/staging/stg_model_5.sql:3]: ugh come on

**Review by dave-connors-3 — APPROVED** [review]
